Comments (#9)
jfy133/taxprofiler-manuscript: taxprofiler-manuscript.qmd

    -r 1.1.0 \
    -profile singularity,<institute> \
    --input <samplesheet.csv> \
    --databases <database.csv> \
    --perform_shortread_qc \
    --shortread_qc_minlength 20 \
    --preprocessing_qc_tool falco \
    --run_host_removal --hostremoval_reference 'host_genome.fasta' \
    --run_kraken2 --kraken2_save_reads \
    --run_metaphlan3 \
    --run_krona \
    --run_profile_standardisation
```

All nf-core pipelines are strictly versioned (specified with the Nextflow `-r` flag), and to ensure reproducibility, each version of the pipeline has a fixed set of software used for each step of the pipeline.
The fixed set of software are controlled through the use of the conda package manager or containers (e.g., Docker, Apptainer [previously known as Singularity]) from the stable Bioconda [@Gruning2018-vr] or BioContainers [@Da_Veiga_Leprevost2017-gl] repositories.
This, coupled with the intrinsic Nextflow ability to execute on most infrastructure whether that is a local laptop (resource requirements permitting), traditional HPC, as well across common cloud providers also makes nf-core/taxprofiler a very portable pipeline that can be used in many contexts.

### Preprocessing

> Preprocessing steps in nf-core/taxprofiler are aimed at removing laboratory and sequencing artefacts that may influence taxonomic profiling, either for computing resource consumption or binary classification reasons.

Reviewer: ```suggestion
Preprocessing steps in nf-core/taxprofiler are aimed at removing laboratory and sequencing artefacts that may influence taxonomic profiling, either for computing resource consumption or and/or false-positive or false-negative classification reasons.
```

I know what I wrote was ugly but I don't have a better way of describing this atm (I wouldn't know what you meant here by 'binary classification')

First sequencing quality control with FastQC [@Andrews2010-pd] or Falco [@De_Sena_Brandine2021-pi] is carried out.
Falco was included for reduced memory requirements, in particular for long read sequencing data.

> Artificial library adapter sequences added during sequencing reduce alignment accuracy by reducing sequence specificity, and in some cases, may result in false-positive hits due to adapter sequence contamination in reference genomes [@Schaffer2018-be;@Breitwieser2018-xg] [^1].

Reviewer: ```suggestion
Artificial library adapter sequences added during sequencing reduce sequencing matching accuracy by reducing sequence specificity, and in some cases, may result in false-positive hits due to adapter sequence contamination in reference genomes [@Schaffer2018-be;@Breitwieser2018-xg] [^1].
```

For these tasks nf-core/taxprofiler can apply either fastp [@Chen2018-vg] or AdapterRemoval2 [@Schubert2016-qv] for short reads, and currently Porechop [@Wick2017-cn] for Oxford Nanopore long-read data.
For both short and long reads, FastQC or Falco is run again to allow assessment on the performance of the adapter removal and/or pair-merging step.

Low complexity sequences, e.g. sequences containing long stretches of mono- or di-nucleotide repeats provide little specific genetic information that contribute to taxonomic identification, as they can align to many different reference genomes [@Schmieder2011-ea;@Clarke2019-al].
Including such reads during taxonomic profiling can increase run-time and memory usage for little gain, as during lowest-common-ancestor (LCA) classification steps they will be assigned to high-level taxonomic ranks (e.g. Kingdom).
nf-core/taxprofiler performs removal of these reads through complexity filtering algorithms as provided by fastp, BBDuk [@Bushnell2022-pf], or PRINSEQ++ [@Cantu2019-lh].

> Long-read sequences often do not have such reads, as lengths are sufficient enough to capture greater sequence diversity - but it is sometimes desirable to only classify reads longer than a certain length - as these provide more precise taxonomic information [@Dilthey2019-ps;@Portik2022-jp].

Reviewer: ```suggestion
Long read sequences often do not have such reads, as lengths are sufficient enough to capture greater sequence diversity - but it is sometimes desirable to only classify reads longer than a certain length - as these provide more precise taxonomic information [@Dilthey2019-ps;@Portik2022-jp].
```

Therefore, nf-core/taxprofiler can remove reads shorter than a user-defined length using Filtlong.

Removing host DNA is another common preprocessing step in metagenomic studies.
This can help speed up run-time, particularly in microbiome studies, where detection of microbes are of interest.
Furthermore, host-contamination of reference genomes in public databases is common [@Longo2011-qd;@Kryukov2016-my;@Breitwieser2019-iz] and therefore the removal of such sequences can also decrease the risk of false positive taxonomic assignment.
To remove multiple hosts or other sequences, all reference genomes can be combined into a single FASTA reference file.
Short read host removal can be carried out with Bowtie2 [@Langmead2012-ik;@Langmead2019-ej] and minimap2 [@Li2018-qq] for long reads, both in combination with SAMtools [@Li2009-wy;@Danecek2021-gj], where reads are aligned against the reference genome and the off-target (unaligned) reads are then converted back to FASTQ format for classification.

Finally, nf-core/taxprofiler can optionally perform run merging where libraries have been sequenced over multiple lanes to generate one profile per sample or library.

> The final set of reads used for profiling can be optionally saved for downstream re-use. Throughout all steps, relevant statistics and log files are generated and used both for the final pipeline run report as well as saved into the results directory of the pipeline run for further inspection where necessary.

Reviewer: ```suggestion
The final set of reads used for profiling can be optionally saved for downstream re-use.
Throughout all steps, relevant statistics and log files are generated and used both for the final pipeline run report as well as saved into the results directory of the pipeline run for further inspection where necessary.
```

### Profiling

There are many types of metagenomic profiling techniques, from profiling against whole-genome references with alignment or k-mer based approaches, to methods involving alignment to species-specific marker-gene families [@Quince2017-bl;@Ye2019-fl].
nf-core/taxprofiler aims to support and include all established classification or profiling tools as requested by the community.

> <!--# (Moritz): I was actually wondering about this the other day. I think there are two possible approaches: 1. We add all tools on a per request basis as stated here. Upside: Everybody's happy. Downside: Maintenance becomes ever more burdensome over time. Do we ever remove profilers? 2. The taxprofiler maintainers choose the "best" representative tool within a specific category. -->

Reviewer 2: I think it would be good to be able to remove profilers in the future.

Reviewer: Why would maintainence become more burdensome over time? They should be fixed with the containers etc., no?

I would just keep adding as people ask. I 'm not sure we would have enough expertise to pick the 'best' representative tool for every discipline...

Author: You have more and more modules whose versions need to be updated and potentially arguments need to be fixed. Additionally, the pipeline logic itself keeps growing and becomes harder to grok.

However, users are still welcome to supply their own custom configuration files, further refining computational limitations or execution specifications.

### Post-profiling

In metagenomic studies, it is common practise to compare the profiles among many samples, and the results of multiple profiles are normally stored in 'taxon tables', i.e, counts per reference taxon (rows), for each sample (columns).
When available, nf-core/taxprofiler supports the option to produce the 'native' taxon table of each classification tool when multiple samples are run.

One of the challenges that researchers face when comparing multiple taxonomic classifiers or profilers is the heterogenous output formats that are produced, that often require custom parsing and merging scripts for each tool to standardise.
To facilitate more user-friendly cross-comparisons between tools, nf-core/taxprofiler utilises the TAXPASTA tool [@Beber2023-hk] to generate standardised profiles and generate multi-sample tables.

> Summary statistics for the entire pipeline are visualised and displayed in a customisable MultiQC report [@Ewels2020-vi]. When supported, quality control of data and pipeline runs are shown for manual verification.

Reviewer: ```suggestion
Summary statistics for the entire pipeline are visualised and displayed in a customisable MultiQC report [@Ewels2020-vi].
When supported, quality control of data and pipeline runs are shown for manual verification.
```

### Comparison with other solutions

nf-core/taxprofiler has been specifically developed for the analysis of whole-genome, _metagenomic_ sequencing data.

> While other types of taxonomic profiling data such as 16S amplicon sequencing are well established fields with a range of popular high-quality and best-practise tools pipelines (e.g. [@Blanco-Miguez2023-cq;@Schloss2009-vk]) and databases [@DeSantis2006-cm;@Yilmaz2014-js], 'gold standard' tools and databases for metagenomics remain much less established. Thus, the need for highly-multiplexed classification is more desirable for the newer metagenomics methods.

Reviewer: ```suggestion
While other types of taxonomic profiling data such as 16S amplicon sequencing are well established fields with a range of popular high-quality and best-practise tools pipelines (e.g. [@Blanco-Miguez2023-cq;@Schloss2009-vk]) and databases [@DeSantis2006-cm;@Yilmaz2014-js], 'gold standard' tools and databases for metagenomics remain much less established.
Thus, the need for highly-multiplexed classification is more desirable for the newer metagenomics methods.
```

However, there are varying levels of detail and comprehensiveness. In particular, StaG-mwc and nf-core/taxprofiler have the most detailed descriptions of all possible output files for every supported module, whereas Unipro UGENE and sunbeam have very minimal to possibly unfinished output documentation.
For execution options, most of the pipelines provide CLI execution, except for Unipro UGENE which offers only GUI-based pipeline set-up (despite a command-line execution of the GUI generated configuration). In particular, nf-core/taxprofiler is the only pipeline providing both CLI and GUI interfaces for pipeline run execution.

Criteria covering portability also overlap with accessibility, as it implies options for and ease of different users running on different types of computing infrastructure, whether that is on their own laptop, on an HPC cluster, or in the cloud.
Unipro UGENE is the only pipeline that supports execution on all three major operating systems (Linux, OSX, Windows), whereas StaG-mwc and nf-core/taxprofiler can be run on unix operating systems, and sunbeam and TAMA are only being supported on Linux.

> <!--# (Moritz): Is this so? Nextflow requires bash, but there is git bash, anaconda terminal, and the WSL for Windows. So this seems a bit strict? -->

Reviewer: Does WSL work reliably though and that is sort of technically still 'linux' (even if _within_ Windows)? I've seen varying reports... I was basing it on what had been reported

Author: I'm not a user of the WSL myself, so I don't know.

In terms of reproducibility, all five pipelines are good at ensuring reproducibility in terms of pipeline and software versioning (allowing re-execution of pipeline runs using the same software), with only tama not having stable versioned releases.
However, installing software manually across different infrastructures can result in variability in the execution of each software [^2] [@Di_Tommaso2017-xu].The current most popular solution to the problem of inconsistent software environments is to use container engines such as Docker or Apptainer to run container images which are isolated, deterministic computing environments which can be executed by any system providing a container runtime.

> Only Unipro UGENE does not document the use of a container system with nf-core/taxprofiler offering the biggest choice for users courtesy of Nextflow (6 different engine systems at the time of writing).

Reviewer: ```suggestion
Only Unipro UGENE does not document the use of a container system, with nf-core/taxprofiler offering the biggest choice for users courtesy of Nextflow (6 different engine systems at the time of writing).
```

In addition to the more precise classification, post-classification read-alignment could also be particularly useful for researchers in palaeogenomics who wish to use tools other than KrakenUniq for initial classification (as in aMeta), where alignment information can be used to authenticate ancient DNA within their samples but also in clinical metagenomics to identify potential pathogens at much finer resolution (e.g. down to strain level).

<!-- TODO additional post-processing steps? e.g. decontamination estimates? -->

Another motivation for developing nf-core/taxprofiler, despite the large number of existing metagenomics pipelines is that by establishing a taxonomic profiling pipeline within the nf-core ecosystem, it is possible to begin building both standalone but also an integrated suite of powerful interconnected pipelines for the major stages of metagenomic workflows.
Existing microbial- and metagenomics- related pipelines within the nf-core initiative include nf-core/ampliseq, nf-core/mag, and nf-core/funcscan.

> We expect over time the ability to link inputs and outputs of each workflow to develop comprehensive metagenomic analyses, while still maintaing powerful standalone pipelines, providing maximal user choice.

Reviewer: ```suggestion
We expect over time the ability to link inputs and outputs of each workflow to develop comprehensive metagenomic analyses, while still maintaining powerful standalone pipelines, providing maximal user choice.
```
